File: integration/ts/sphincs-wallet.test.ts
/**
 * SPHINCSWallet integration test — Hybrid ECDSA + SLH-DSA-SHA2-128s contract.
 *
 * ## Security Model: Two-Layer Authentication
 *
 * This contract creates a quantum-resistant spending path by combining
 * classical ECDSA with SLH-DSA (FIPS 205, SPHINCS+):
 *
 * 1. **ECDSA** proves the signature commits to this specific transaction
 *    (via OP_CHECKSIG over the sighash preimage).
 * 2. **SLH-DSA** proves the ECDSA signature was authorized by the SLH-DSA
 *    key holder — the ECDSA signature bytes ARE the message that SLH-DSA signs.
 *
 * A quantum attacker who can break ECDSA could forge a valid ECDSA
 * signature, but they cannot produce a valid SLH-DSA signature over their
 * forged sig without knowing the SLH-DSA secret key. SLH-DSA security
 * relies only on SHA-256 collision resistance, not on any number-theoretic
 * assumption vulnerable to Shor's algorithm.
 *
 * Unlike WOTS+ (one-time), SLH-DSA is stateless and the same keypair
 * can sign many messages — it's NIST FIPS 205 standardized.
 *
 * ## Constructor
 *   - ecdsaPubKeyHash: Addr — 20-byte HASH160 of compressed ECDSA public key
 *   - slhdsaPubKeyHash: ByteString — 20-byte HASH160 of 32-byte SLH-DSA public key
 *
 * ## Method: spend(slhdsaSig, slhdsaPubKey, sig, pubKey)
 *   - slhdsaSig: 7,856-byte SLH-DSA-SHA2-128s signature
 *   - slhdsaPubKey: 32-byte SLH-DSA public key (PK.seed[16] || PK.root[16])
 *   - sig: ~72-byte DER-encoded ECDSA signature + sighash flag
 *   - pubKey: 33-byte compressed ECDSA public key
 *
 * ## Script Size
 *   ~188 KB — SLH-DSA verification requires computing multiple WOTS+
 *   verifications and Merkle tree path checks within the Bitcoin Script VM.
 *
 * ## Test Approach
 *   Deployment tests use hash commitments of test keys. Full spending tests
 *   require raw transaction construction (two-pass signing: ECDSA first, then
 *   SLH-DSA over the ECDSA sig). The Go integration suite (TestSLHDSA_ValidSpend)
 *   implements the complete two-pass spending flow.
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract, RPCProvider } from 'runar-sdk';
import { createFundedWallet } from './helpers/wallet.js';
import { createHash } from 'node:crypto';

function createProvider() {
  return new RPCProvider('http://localhost:18332', 'bitcoin', 'bitcoin', {
    autoMine: true,
    network: 'testnet',
  });
}

function hash160hex(data: Buffer | Uint8Array): string {
  const sha = createHash('sha256').update(data).digest();
  return createHash('ripemd160').update(sha).digest('hex');
}

// Deterministic SLH-DSA test public key (32 bytes hex: PK.seed[16] || PK.root[16])
// Generated from seed [0, 1, 2, ..., 47] with SLH-DSA-SHA2-128s (n=16).
const SLHDSA_TEST_PK = '00000000000000000000000000000000b618cb38f7f785488c9768f3a2972baf';
const SLHDSA_TEST_PK_HASH = hash160hex(Buffer.from(SLHDSA_TEST_PK, 'hex'));

describe('SPHINCSWallet (Hybrid ECDSA + SLH-DSA-SHA2-128s)', () => {
  it('should compile the contract', () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('SPHINCSWallet');
    expect(artifact.script.length).toBeGreaterThan(0);
  });

  it('should produce a very large script (~188 KB)', () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');
    const scriptBytes = artifact.script.length / 2;
    // SLH-DSA scripts are typically ~188 KB
    expect(scriptBytes).toBeGreaterThan(100000);
    expect(scriptBytes).toBeLessThan(500000);
  });

  it('should deploy with ECDSA + SLH-DSA keys', async () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    // Constructor: (ecdsaPubKeyHash, slhdsaPubKeyHash)
    const contract = new RunarContract(artifact, [pubKeyHash, SLHDSA_TEST_PK_HASH]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 50000 });
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should deploy with a different SLH-DSA public key', async () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    // Different SLH-DSA public key
    const otherPK = 'aabbccdd00000000000000000000000011223344556677889900aabbccddeeff';
    const otherPKHash = hash160hex(Buffer.from(otherPK, 'hex'));
    const contract = new RunarContract(artifact, [pubKeyHash, otherPKHash]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 50000 });
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy and verify UTXO exists (spend requires raw tx construction)', async () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    // Deploy the hybrid ECDSA+SLH-DSA contract
    const contract = new RunarContract(artifact, [pubKeyHash, SLHDSA_TEST_PK_HASH]);
    await contract.deploy(provider, signer, { satoshis: 50000 });

    // The hybrid spend pattern requires:
    //   1. Build unsigned spending transaction
    //   2. ECDSA-sign the transaction input
    //   3. SLH-DSA-sign the ECDSA signature bytes
    //   4. Construct unlocking script: <slhdsaSig> <slhdsaPK> <ecdsaSig> <ecdsaPubKey>
    //
    // This two-pass signing pattern is fully tested in the Go integration suite
    // (TestSLHDSA_ValidSpend) which uses raw transaction construction.
    expect(contract.getUtxo()).toBeTruthy();
  });

  it('should deploy with tampered key hash (for rejection testing)', async () => {
    const artifact = compileContract('examples/ts/sphincs-wallet/SPHINCSWallet.runar.ts');

    const provider = createProvider();
    const { signer, pubKeyHash } = await createFundedWallet(provider);

    // Deploy succeeds (deployment doesn't verify the hash, that happens at spend time)
    const contract = new RunarContract(artifact, [pubKeyHash, SLHDSA_TEST_PK_HASH]);
    await contract.deploy(provider, signer, { satoshis: 50000 });

    // Contract deployed with correct hash commitments; tampered spend would fail
    // at the OP_HASH160 <slhdsaPubKeyHash> OP_EQUALVERIFY step if a wrong key is provided.
    expect(contract.getUtxo()).toBeTruthy();
  });
});

File: integration/zig/src/sphincs_wallet_test.zig
const std = @import("std");
const runar = @import("runar");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

test "SPHINCSWallet_Compile" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig") catch |err| {
        std.log.warn("Could not compile SPHINCSWallet contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("SPHINCSWallet", artifact.contract_name);
    std.log.info("SPHINCSWallet compiled: {d} bytes", .{artifact.script.len / 2});
}

test "SPHINCSWallet_ScriptSize" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig") catch |err| {
        std.log.warn("Could not compile SPHINCSWallet contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    const script_bytes = artifact.script.len / 2;
    // Hybrid ECDSA+SLH-DSA contracts produce very large scripts (100-500 KB)
    try std.testing.expect(script_bytes >= 100000);
    try std.testing.expect(script_bytes <= 500000);
    std.log.info("SPHINCSWallet script size: {d} bytes", .{script_bytes});
}

test "SPHINCSWallet_Deploy" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig") catch |err| {
        std.log.warn("Could not compile SPHINCSWallet contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var ecdsa_wallet = try helpers.newWallet(allocator);
    defer ecdsa_wallet.deinit();

    const ecdsa_pkh = try ecdsa_wallet.pubKeyHashHex(allocator);
    defer allocator.free(ecdsa_pkh);

    // SLH-DSA public key hash: use a deterministic 20-byte (40-char hex) hash
    // In production this would be hash160(SLH-DSA public key)
    const slhdsa_pkh_hex = "0102030405060708091011121314151617181920";

    // Constructor: ecdsaPubKeyHash, slhdsaPubKeyHash
    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = ecdsa_pkh },
        .{ .bytes = slhdsa_pkh_hex },
    });
    defer contract.deinit();

    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 50000 });
    defer allocator.free(deploy_txid);

    const utxo = contract.getCurrentUtxo();
    try std.testing.expect(utxo != null);
    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("SPHINCSWallet deployed: {s}", .{deploy_txid});
}

test "SPHINCSWallet_DeployDifferentKey" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig") catch |err| {
        std.log.warn("Could not compile SPHINCSWallet contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var ecdsa_wallet = try helpers.newWallet(allocator);
    defer ecdsa_wallet.deinit();
    const ecdsa_pkh = try ecdsa_wallet.pubKeyHashHex(allocator);
    defer allocator.free(ecdsa_pkh);

    // Keypair 1
    const pkh1 = "0102030405060708091011121314151617181920";

    // Keypair 2 (different seed -> different hash)
    const pkh2 = "aabbccddeeff00112233445566778899aabbccdd";

    // Deploy with kp1
    var contract1 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = ecdsa_pkh },
        .{ .bytes = pkh1 },
    });
    defer contract1.deinit();

    var funder1 = try helpers.newWallet(allocator);
    defer funder1.deinit();
    const fund1 = try helpers.fundWallet(allocator, &funder1, 1.0);
    defer allocator.free(fund1);

    var rpc1 = helpers.RPCProvider.init(allocator);
    var signer1 = try funder1.localSigner();
    const txid1 = try contract1.deploy(rpc1.provider(), signer1.signer(), .{ .satoshis = 50000 });
    defer allocator.free(txid1);

    // Deploy with kp2
    var contract2 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = ecdsa_pkh },
        .{ .bytes = pkh2 },
    });
    defer contract2.deinit();

    var funder2 = try helpers.newWallet(allocator);
    defer funder2.deinit();
    const fund2 = try helpers.fundWallet(allocator, &funder2, 1.0);
    defer allocator.free(fund2);

    var rpc2 = helpers.RPCProvider.init(allocator);
    var signer2 = try funder2.localSigner();
    const txid2 = try contract2.deploy(rpc2.provider(), signer2.signer(), .{ .satoshis = 50000 });
    defer allocator.free(txid2);

    try std.testing.expect(!std.mem.eql(u8, txid1, txid2));
    std.log.info("seed1 txid: {s}, seed2 txid: {s}", .{ txid1, txid2 });
}

test "SPHINCSWallet_NotStateful" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/sphincs-wallet/SPHINCSWallet.runar.zig") catch |err| {
        std.log.warn("Could not compile SPHINCSWallet contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // SPHINCSWallet is a stateless contract (extends SmartContract)
    try std.testing.expect(!artifact.isStateful());
}
